boot/freeldr/freeldr/arch/i386/multiboot.S

```diff
@@ -98,73 +100,70 @@ MultibootEntry:
     /* Save multiboot info structure */
     mov esi, ebx
     mov edi, offset MultibootInfo + INITIAL_BASE - FREELDR_BASE
-    mov ecx, (MB_INFO_SIZE / 4)
-    rep movsd
-    mov dword ptr ds:[MultibootInfo + INITIAL_BASE - FREELDR_BASE + MB_INFO_MMAP_ADDR_OFFSET], 0
-    mov dword ptr ds:[_MultibootInfoPtr + INITIAL_BASE - FREELDR_BASE], offset MultibootInfo
+    mov ecx, MB_INFO_SIZE
+    rep movsb
+    mov dword ptr [MultibootInfo + INITIAL_BASE - FREELDR_BASE + MB_INFO_MMAP_ADDR_OFFSET], 0
+    mov dword ptr [_MultibootInfoPtr + INITIAL_BASE - FREELDR_BASE], offset MultibootInfo
 
     /* See if the memory map was passed in */
-    test dword ptr ds:[ebx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_MEMORY_MAP
+    test dword ptr [rbx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_MEMORY_MAP
     jz mbchk_command_line
     /* Check memory map length */
-    mov ecx, dword ptr ds:[ebx + MB_INFO_MMAP_LEN_OFFSET]
+    mov rcx, [rbx + MB_INFO_MMAP_LEN_OFFSET]
     test ecx, ecx
     jz mbchk_command_line
     cmp ecx, MB_MMAP_SIZE
     jg mbchk_command_line
     /* Check memory map address */
-    mov esi, dword ptr ds:[ebx + MB_INFO_MMAP_ADDR_OFFSET]
+    mov rsi, [rbx + MB_INFO_MMAP_ADDR_OFFSET]
     test esi, esi
     jz mbchk_command_line
     /* Save memory map structure */
     mov edi, offset MultibootMemoryMap + INITIAL_BASE - FREELDR_BASE
-    shr ecx, 2
-    rep movsd
+    rep movsb
     /* Relocate memory map address */
-    mov dword ptr ds:[MultibootInfo + INITIAL_BASE - FREELDR_BASE + MB_INFO_MMAP_ADDR_OFFSET], offset MultibootMemoryMap
+    mov dword ptr [MultibootInfo + INITIAL_BASE - FREELDR_BASE + MB_INFO_MMAP_ADDR_OFFSET], offset MultibootMemoryMap
 
 mbchk_command_line:
     /* Save command line */
-    test dword ptr ds:[ebx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_COMMAND_LINE
+    test dword ptr [rbx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_COMMAND_LINE
     jz mb2
-    mov esi, dword ptr ds:[ebx + MB_INFO_COMMAND_LINE_OFFSET]
+    mov rsi, [rbx + MB_INFO_COMMAND_LINE_OFFSET]
     mov edi, offset cmdline + INITIAL_BASE - FREELDR_BASE
     mov ecx, CMDLINE_SIZE - 1
 mb1:
     lodsb
     stosb
     test al, al
     jz mb2
-    dec ecx
+    sub ecx, 1
     jnz mb1
 
 mb2:
     /* See if the boot device was passed in */
-    test dword ptr ds:[ebx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_BOOT_DEVICE
+    test dword ptr [rbx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_BOOT_DEVICE
 
     /* If no boot device known, assume first partition of first harddisk */
     mov dx, HEX(0180)
     jz mb3
 
     /* Load boot drive into DL, boot partition into DH */
-    mov edx, dword ptr ds:[ebx + MB_INFO_BOOT_DEVICE_OFFSET]
+    mov rdx, [rbx + MB_INFO_BOOT_DEVICE_OFFSET]
     bswap edx
-    inc dh
```

Comment:
I think this one is to keep with our own internal conventions (partition number being 1-based).

Reply:
but this breaks the cdrom boot via grub. and i tested in my Hard Drive and it works without `inc dh`.

Reply:
Yes. apparently.

Reply:
According to the multiboot spec, and _**if I understand it correctly**_, this is not so:
https://www.gnu.org/software/grub/manual/multiboot/multiboot.html
```
If bit 1 in the ‘flags’ word is set, then the ‘boot_device’ field is valid,
and indicates which BIOS disk device the boot loader loaded the OS image from.
If the OS image was not loaded from a BIOS disk, then this field must not be
present (bit 3 must be clear). The operating system may use this field as a
hint for determining its own root device, but is not required to. The
‘boot_device’ field is laid out in four one-byte subfields as follows:

+-------+-------+-------+-------+
| part3 | part2 | part1 | drive |
+-------+-------+-------+-------+
Least significant             Most significant

The most significant byte contains the BIOS drive number as understood by the
BIOS INT 0x13 low-level disk interface: e.g. 0x00 for the first floppy disk
or 0x80 for the first hard disk.

The three remaining bytes specify the boot partition. ‘part1’ specifies the
top-level partition number, ‘part2’ specifies a sub-partition in the top-level
partition, etc. Partition numbers always start from zero. Unused partition
bytes must be set to 0xFF. For example, if the disk is partitioned using a
simple one-level DOS partitioning scheme, then ‘part1’ contains the DOS
partition number, and ‘part2’ and ‘part3’ are both 0xFF. As another example,
if a disk is partitioned first into DOS partitions, and then one of those
DOS partitions is subdivided into several BSD partitions using BSD’s disklabel
strategy, then ‘part1’ contains the DOS partition number, ‘part2’ contains
the BSD sub-partition within that DOS partition, and ‘part3’ is 0xFF.

DOS extended partitions are indicated as partition numbers starting from 4 and
increasing, rather than as nested sub-partitions, even though the underlying
disk layout of extended partitions is hierarchical in nature. For example, if
the boot loader boots from the second extended partition on a disk partitioned
in conventional DOS style, then ‘part1’ will be 5, and ‘part2’ and ‘part3’
will both be 0xFF.
```

Reply:
ah, yes the partitions numbers start from zero.

Reply:
and the freeldr uses the partition number from `dh`?

Reply:
Actually it doesn't that much (see usage of `FrLdrBootPartition`).
It mostly checks whether it's != (or ==) 0xFF, for cdroms.

Reply:
so `inc dh` was just causing problems

Reply:
i added back the `inc dh`